#!/usr/bin/Rscript

# takes a "nearest" csv file, as generated by mm9RDtoGenes.R and
# generates some useful stats and plots from it

options(stringsAsFactors = FALSE);

library(IRanges)

qw <- function(...) {
  as.character(sys.call()[-1])
}

args <- commandArgs(trailingOnly=TRUE)
filename = args[1]
dat <- read.csv(filename)



#plot for all and then seperately by type
types <- unique(dat[,"type"])
if(length(types)>1){types <- c("all",types)}

for (ty in types){
  if(ty=="all"){
    this.dat <- dat
  }else{
    this.dat <- dat[dat[,"type"]==ty,]
  }
  i <-  this.dat$insideFeature
  overlapping.transcript <- i == "inside" | i == "overlapStart" | i == "overlapEnd" | i=="includeFeature"
  exonic <- this.dat$overlapping.exon == 1
  intronic <- overlapping.transcript &! exonic
  
  test <- cbind(overlapping.transcript, exonic, intronic) 
  test <- which(apply(test, 1,sum)==1)
  other.exonic = rep(0, nrow(this.dat))
  other.exonic[test] <- 1
  exonic[test] <- 0
  
  nms <- c( sum(overlapping.transcript),
           nrow(this.dat)-sum(overlapping.transcript))
  
  labels <-  c(paste("overlapping gene (", sum(overlapping.transcript) ,")", sep=""),
               paste("not overlapping gene (",nrow(this.dat)-sum(overlapping.transcript) ,")", sep=""))
  
  my.cols <- colors()[c(5,2)]
  title = ""
  
  p.file <- sub('.csv$',paste('_',ty,'_ingenes_pie.ps', sep=""), filename)
  postscript(p.file)
  pie(nms,labels=labels, col=my.cols, main=title) 
  dev.off()
  
  p.file <- sub('.csv$',paste('_',ty,'_ingenes_box.ps', sep=""), filename)
  postscript(p.file)
  barplot(nms, names.arg=labels, col=my.cols, main=title)
  dev.off()
  
  
                                        # of stuff in a gene region, what's exonic, what's intronic? (with reference to the nearest gene, so
                                        # stuff annotate as not-in-a-gene could still be in the exon/intron of another gene)
  nms <- c( sum(exonic),
           sum(intronic))
  
  labels <-  c(paste("exonic (", sum(exonic) ,")", sep=""),
               paste("intronic (",sum(intronic) ,")", sep=""))
  
  my.cols <- colors()[c(5,2)]
  title = ""
  
  p.file <- sub('.csv$',paste('_',ty,'_exonintron_pie.ps', sep=""), filename)
  postscript(p.file)
  pie(nms,labels=labels, col=my.cols, main=title)
  dev.off()
  
  p.file <- sub('.csv$',paste('_',ty,'_exonintron_box.ps', sep=""), filename)
  postscript(p.file)
  barplot(nms, names.arg=labels, col=my.cols, main=title)
  dev.off()
  
  
                                        #density distributions and such
  this.d <- this.dat$distancetoFeature
  this.ad <- abs(this.d)
  
  p.file <- sub('.csv$',paste('_',ty,'_densitydist.ps', sep=""), filename)
  postscript(p.file)
  plot(density(this.d), main="Density Distribution of Distances to Nearest Peak")
  dev.off()
  
  p.file <- sub('.csv$',paste('_',ty,'_densitydist_within100k.ps', sep=""), filename)
  postscript(p.file)
  plot(density(this.d[this.ad<100000]), main="Density Distribution of Distances to Nearest Peak (within 100kb")
  dev.off()
  
  
  p.file <- sub('.csv$',paste('_',ty,'_distvFoldEnrich.ps', sep=""), filename)
  postscript(p.file)
  plot(this.d, this.dat[,"FoldEnrichment"], pch="+", main="FoldEnrichment vs Distance to Nearest Peak")
  dev.off()
  
  p.file <- sub('.csv$',paste('_',ty,'_distvFoldEnrich_within100k.ps', sep=""), filename)
  postscript(p.file)
  plot(this.d[this.ad<100000], this.dat[(this.ad<100000),"FoldEnrichment"], pch="+", main="FoldEnrichment vs Distance to Nearest Peak")
  dev.off()
  
  
  p.file <- sub('.csv$',paste('_',ty,'_neg10log10pVal.ps', sep=""), filename)
  postscript(p.file)
  plot(this.d, this.dat[,"neg10log10pVal"], pch="+", main="-10log10p.val vs Distance to Nearest Peak")
  dev.off()
  
  p.file <- sub('.csv$',paste('_',ty,'_neg10log10pVal_within100k.ps', sep=""), filename)
  postscript(p.file)
  plot(this.d[this.ad<100000], this.dat[(this.ad<100000),"neg10log10pVal"], pch="+", main="-10log10p.val vs Distance to Nearest Peak")
  dev.off()
  
  
  
                                        #because apparently everyone loves an uninformative pie chart:
  within.1kb <- (this.ad < 1000)
  between.1kb.and.5kb <- (this.ad >= 1000 & this.ad < 5000)
  between.5kb.and.10kb <- (this.ad >= 5000 & this.ad < 10000)
  between.10kb.and.100kb <- (this.ad >= 10000 & this.ad < 100000)
  more.than.100kb <- (this.ad >= 100000)
  
  p.file <- sub('.csv$',paste('_',ty,'_distances_pie.ps', sep=""), filename)
  postscript(p.file)
  pie(c("1kb" = sum(within.1kb),
        "1-5kb" = sum(between.1kb.and.5kb),
        "5-10kb" = sum(between.5kb.and.10kb),
        "10-100kb" = sum(between.10kb.and.100kb),
        ">100kb" = sum(more.than.100kb)
        )
      )
  dev.off()

}
